File: src/tools/audio.ts
import { existsSync, copyFileSync, mkdirSync } from "fs";
import { join, basename, extname } from "path";
import { getDatabase, getAudiobookDir, getCurrentProjectPath } from "../db/connection.js";
import type { Segment } from "../db/schema.js";
import { getChapter, listChapters, getChaptersWithStats } from "./chapters.js";
import { listSegments, getSegment, getPendingSegments } from "./segments.js";
import { getProjectInfo } from "./projects.js";
import {
  checkFfmpeg,
  validateAudioFile,
  getAudioDuration,
  concatenateAudioFiles,
  createAudiobookWithChapters,
  type ChapterMarker,
} from "../utils/ffmpeg.js";

export interface RegisterSegmentAudioParams {
  segment_id: string;
  audio_path: string;
  duration_ms?: number;
}

export interface StitchChapterParams {
  chapter_id: string;
  output_filename?: string;
}

export interface StitchBookParams {
  output_filename?: string;
  include_chapter_markers?: boolean;
}

/**
 * Register an audio file for a segment
 */
export function registerSegmentAudio(params: RegisterSegmentAudioParams): {
  segment_id: string;
  audio_path: string;
  duration_ms: number;
  copied_to: string;
} {
  const { segment_id, audio_path, duration_ms } = params;
  const db = getDatabase();
  const projectPath = getCurrentProjectPath();

  if (!projectPath) {
    throw new Error("No project is currently open");
  }

  // Verify segment exists
  const segment = getSegment(segment_id);
  if (!segment) {
    throw new Error(`Segment not found: ${segment_id}`);
  }

  // Validate audio file
  const validation = validateAudioFile(audio_path);
  if (!validation.valid) {
    throw new Error(`Invalid audio file: ${validation.error}`);
  }

  // Use provided duration or detected duration
  const actualDuration = duration_ms ?? validation.duration_ms!;

  // Copy audio to project's audio directory
  const audiobookDir = getAudiobookDir(projectPath);
  const segmentsDir = join(audiobookDir, "audio", "segments");
  if (!existsSync(segmentsDir)) {
    mkdirSync(segmentsDir, { recursive: true });
  }

  const ext = extname(audio_path) || ".mp3";
  const destPath = join(segmentsDir, `${segment_id}${ext}`);
  copyFileSync(audio_path, destPath);

  // Store relative path in database
  const relativePath = `audio/segments/${segment_id}${ext}`;

  // Update segment with audio info
  const stmt = db.prepare(`
    UPDATE segments SET audio_path = ?, duration_ms = ? WHERE id = ?
  `);
  stmt.run(relativePath, actualDuration, segment_id);

  return {
    segment_id,
    audio_path: relativePath,
    duration_ms: actualDuration,
    copied_to: destPath,
  };
}

/**
 * Get the status of audio for a chapter
 */
export function getChapterAudioStatus(chapter_id: string): {
  chapter_id: string;
  chapter_title: string;
  total_segments: number;
  segments_with_audio: number;
  segments_missing_audio: number;
  total_duration_ms: number;
  ready_to_stitch: boolean;
  missing_segments: Array<{
    id: string;
    sort_order: number;
    text_preview: string;
    character_name: string | null;
  }>;
} {
  const db = getDatabase();

  // Verify chapter exists
  const chapter = getChapter(chapter_id);
  if (!chapter) {
    throw new Error(`Chapter not found: ${chapter_id}`);
  }

  const segments = listSegments(chapter_id);
  const withAudio = segments.filter((s) => s.audio_path);
  const missingAudio = segments.filter((s) => !s.audio_path);

  // Get character names for missing segments
  const missingWithNames = missingAudio.map((s) => {
    let characterName: string | null = null;
    if (s.character_id) {
      const charResult = db
        .prepare(`SELECT name FROM characters WHERE id = ?`)
        .get(s.character_id) as { name: string } | undefined;
      characterName = charResult?.name ?? null;
    }
    return {
      id: s.id,
      sort_order: s.sort_order,
      text_preview:
        s.text_content.length > 50
          ? s.text_content.slice(0, 50) + "..."
          : s.text_content,
      character_name: characterName,
    };
  });

  const totalDuration = withAudio.reduce(
    (sum, s) => sum + (s.duration_ms ?? 0),
    0
  );

  return {
    chapter_id,
    chapter_title: chapter.title,
    total_segments: segments.length,
    segments_with_audio: withAudio.length,
    segments_missing_audio: missingAudio.length,
    total_duration_ms: totalDuration,
    ready_to_stitch: missingAudio.length === 0 && segments.length > 0,
    missing_segments: missingWithNames,
  };
}

/**
 * Stitch all segments in a chapter into a single audio file
 */
export function stitchChapter(params: StitchChapterParams): {
  chapter_id: string;
  chapter_title: string;
  output_path: string;
  segment_count: number;
  total_duration_ms: number;
} {
  const { chapter_id, output_filename } = params;
  const projectPath = getCurrentProjectPath();

  if (!projectPath) {
    throw new Error("No project is currently open");
  }

  // Check ffmpeg availability
  if (!checkFfmpeg()) {
    throw new Error("ffmpeg is not installed or not in PATH");
  }

  // Verify chapter exists and has all audio
  const status = getChapterAudioStatus(chapter_id);
  if (!status.ready_to_stitch) {
    throw new Error(
      `Chapter is not ready to stitch. Missing ${status.segments_missing_audio} audio files.`
    );
  }

  const audiobookDir = getAudiobookDir(projectPath);
  const segments = listSegments(chapter_id);

  // Get absolute paths for all audio files
  const inputFiles = segments.map((s) => join(audiobookDir, s.audio_path!));

  // Determine output path
  const filename =
    output_filename ||
    `${status.chapter_title.replace(/[^a-zA-Z0-9]/g, "_")}.mp3`;
  const outputDir = join(audiobookDir, "exports", "chapters");
  if (!existsSync(outputDir)) {
    mkdirSync(outputDir, { recursive: true });
  }
  const outputPath = join(outputDir, filename);

  // Concatenate files
  concatenateAudioFiles(inputFiles, outputPath, "mp3");

  // Get actual duration of output
  const duration = getAudioDuration(outputPath);

  return {
    chapter_id,
    chapter_title: status.chapter_title,
    output_path: outputPath,
    segment_count: segments.length,
    total_duration_ms: duration,
  };
}

/**
 * Get the overall stitch status for the book
 */
export function getStitchStatus(): {
  total_chapters: number;
  chapters_ready: number;
  total_segments: number;
  segments_with_audio: number;
  total_duration_ms: number;
  ready_to_stitch_book: boolean;
  chapters: Array<{
    id: string;
    title: string;
    sort_order: number;
    segment_count: number;
    segments_with_audio: number;
    duration_ms: number;
    ready: boolean;
  }>;
} {
  const chapters = getChaptersWithStats();

  const chapterStatuses = chapters.map((ch) => {
    const status = getChapterAudioStatus(ch.id);
    return {
      id: ch.id,
      title: ch.title,
      sort_order: ch.sort_order,
      segment_count: ch.segment_count,
      segments_with_audio: status.segments_with_audio,
      duration_ms: status.total_duration_ms,
      ready: status.ready_to_stitch,
    };
  });

  const totalSegments = chapterStatuses.reduce(
    (sum, ch) => sum + ch.segment_count,
    0
  );
  const totalWithAudio = chapterStatuses.reduce(
    (sum, ch) => sum + ch.segments_with_audio,
    0
  );
  const totalDuration = chapterStatuses.reduce(
    (sum, ch) => sum + ch.duration_ms,
    0
  );
  const chaptersReady = chapterStatuses.filter((ch) => ch.ready).length;

  return {
    total_chapters: chapters.length,
    chapters_ready: chaptersReady,
    total_segments: totalSegments,
    segments_with_audio: totalWithAudio,
    total_duration_ms: totalDuration,
    ready_to_stitch_book:
      chaptersReady === chapters.length && chapters.length > 0,
    chapters: chapterStatuses,
  };
}

/**
 * Stitch all chapters into a complete audiobook
 */
export function stitchBook(params: StitchBookParams): {
  output_path: string;
  chapter_count: number;
  total_duration_ms: number;
  chapters: Array<{
    title: string;
    start_ms: number;
  }>;
} {
  const { output_filename, include_chapter_markers = true } = params;
  const projectPath = getCurrentProjectPath();

  if (!projectPath) {
    throw new Error("No project is currently open");
  }

  // Check ffmpeg availability
  if (!checkFfmpeg()) {
    throw new Error("ffmpeg is not installed or not in PATH");
  }

  // Check overall status
  const status = getStitchStatus();
  if (!status.ready_to_stitch_book) {
    const notReady = status.chapters.filter((ch) => !ch.ready);
    throw new Error(
      `Book is not ready to stitch. Chapters not ready: ${notReady.map((ch) => ch.title).join(", ")}`
    );
  }

  const audiobookDir = getAudiobookDir(projectPath);
  const info = getProjectInfo();
  const chapters = listChapters();

  // First, stitch each chapter and collect the chapter files
  const chapterFiles: string[] = [];
  const chapterMarkers: ChapterMarker[] = [];
  let currentMs = 0;

  for (const chapter of chapters) {
    // Stitch the chapter
    const result = stitchChapter({ chapter_id: chapter.id });
    chapterFiles.push(result.output_path);

    // Record chapter marker
    chapterMarkers.push({
      title: chapter.title,
      start_ms: currentMs,
    });

    currentMs += result.total_duration_ms;
  }

  // Determine output path
  const filename =
    output_filename ||
    `${info.project.title.replace(/[^a-zA-Z0-9]/g, "_")}.mp3`;
  const outputDir = join(audiobookDir, "exports", "book");
  if (!existsSync(outputDir)) {
    mkdirSync(outputDir, { recursive: true });
  }
  const outputPath = join(outputDir, filename);

  // Create final audiobook
  if (include_chapter_markers) {
    createAudiobookWithChapters(chapterFiles, outputPath, chapterMarkers, {
      title: info.project.title,
      artist: info.project.author ?? undefined,
      album: info.project.title,
    });
  } else {
    concatenateAudioFiles(chapterFiles, outputPath, "mp3");
  }

  // Get actual duration
  const duration = getAudioDuration(outputPath);

  return {
    output_path: outputPath,
    chapter_count: chapters.length,
    total_duration_ms: duration,
    chapters: chapterMarkers,
  };
}

/**
 * Clear audio from a segment
 */
export function clearSegmentAudio(segment_id: string): void {
  const db = getDatabase();

  // Verify segment exists
  const segment = getSegment(segment_id);
  if (!segment) {
    throw new Error(`Segment not found: ${segment_id}`);
  }

  // Clear audio path and duration
  const stmt = db.prepare(`
    UPDATE segments SET audio_path = NULL, duration_ms = NULL WHERE id = ?
  `);
  stmt.run(segment_id);
}

File: audiobook_mcp/tools/audio.py
"""Audio registration and stitching tools."""

import os
import shutil
from pathlib import Path
from dataclasses import dataclass
from typing import Optional

from ..db.connection import get_database, get_audiobook_dir, get_current_project_path
from .chapters import get_chapter, list_chapters, get_chapters_with_stats
from .segments import list_segments, get_segment
from .projects import get_project_info
from ..utils.ffmpeg import (
    check_ffmpeg,
    validate_audio_file,
    get_audio_duration,
    concatenate_audio_files,
    create_audiobook_with_chapters,
    ChapterMarker,
)


@dataclass
class RegisterResult:
    segment_id: str
    audio_path: str
    duration_ms: int
    copied_to: str


@dataclass
class MissingSegment:
    id: str
    sort_order: int
    text_preview: str
    character_name: Optional[str]


@dataclass
class ChapterAudioStatus:
    chapter_id: str
    chapter_title: str
    total_segments: int
    segments_with_audio: int
    segments_missing_audio: int
    total_duration_ms: int
    ready_to_stitch: bool
    missing_segments: list[MissingSegment]


@dataclass
class ChapterStatus:
    id: str
    title: str
    sort_order: int
    segment_count: int
    segments_with_audio: int
    duration_ms: int
    ready: bool


@dataclass
class StitchStatus:
    total_chapters: int
    chapters_ready: int
    total_segments: int
    segments_with_audio: int
    total_duration_ms: int
    ready_to_stitch_book: bool
    chapters: list[ChapterStatus]


@dataclass
class StitchChapterResult:
    chapter_id: str
    chapter_title: str
    output_path: str
    segment_count: int
    total_duration_ms: int


@dataclass
class ChapterTimestamp:
    title: str
    start_ms: int


@dataclass
class StitchBookResult:
    output_path: str
    chapter_count: int
    total_duration_ms: int
    chapters: list[ChapterTimestamp]


def register_segment_audio(
    segment_id: str,
    audio_path: str,
    duration_ms: Optional[int] = None,
) -> RegisterResult:
    """Register an audio file for a segment."""
    db = get_database()
    cursor = db.cursor()
    project_path = get_current_project_path()

    if not project_path:
        raise RuntimeError("No project is currently open")

    # Verify segment exists
    segment = get_segment(segment_id)
    if not segment:
        raise ValueError(f"Segment not found: {segment_id}")

    # Validate audio file
    validation = validate_audio_file(audio_path)
    if not validation.valid:
        raise ValueError(f"Invalid audio file: {validation.error}")

    # Use provided duration or detected duration
    actual_duration = duration_ms if duration_ms is not None else validation.duration_ms

    # Copy audio to project's audio directory
    audiobook_dir = get_audiobook_dir(project_path)
    segments_dir = Path(audiobook_dir) / "audio" / "segments"
    segments_dir.mkdir(parents=True, exist_ok=True)

    ext = Path(audio_path).suffix or ".mp3"
    dest_path = segments_dir / f"{segment_id}{ext}"
    shutil.copy2(audio_path, dest_path)

    # Store relative path in database
    relative_path = f"audio/segments/{segment_id}{ext}"

    # Update segment with audio info
    cursor.execute(
        "UPDATE segments SET audio_path = ?, duration_ms = ? WHERE id = ?",
        (relative_path, actual_duration, segment_id)
    )
    db.commit()

    return RegisterResult(
        segment_id=segment_id,
        audio_path=relative_path,
        duration_ms=actual_duration or 0,
        copied_to=str(dest_path),
    )


def get_chapter_audio_status(chapter_id: str) -> ChapterAudioStatus:
    """Get the status of audio for a chapter."""
    db = get_database()
    cursor = db.cursor()

    # Verify chapter exists
    chapter = get_chapter(chapter_id)
    if not chapter:
        raise ValueError(f"Chapter not found: {chapter_id}")

    segments = list_segments(chapter_id)
    with_audio = [s for s in segments if s.audio_path]
    missing_audio = [s for s in segments if not s.audio_path]

    # Get character names for missing segments
    missing_with_names: list[MissingSegment] = []
    for s in missing_audio:
        character_name: Optional[str] = None
        if s.character_id:
            cursor.execute("SELECT name FROM characters WHERE id = ?", (s.character_id,))
            row = cursor.fetchone()
            if row:
                character_name = row["name"]

        text_preview = (
            s.text_content[:50] + "..."
            if len(s.text_content) > 50
            else s.text_content
        )

        missing_with_names.append(MissingSegment(
            id=s.id,
            sort_order=s.sort_order,
            text_preview=text_preview,
            character_name=character_name,
        ))

    total_duration = sum(s.duration_ms or 0 for s in with_audio)

    return ChapterAudioStatus(
        chapter_id=chapter_id,
        chapter_title=chapter.title,
        total_segments=len(segments),
        segments_with_audio=len(with_audio),
        segments_missing_audio=len(missing_audio),
        total_duration_ms=total_duration,
        ready_to_stitch=len(missing_audio) == 0 and len(segments) > 0,
        missing_segments=missing_with_names,
    )


def stitch_chapter(
    chapter_id: str,
    output_filename: Optional[str] = None,
) -> StitchChapterResult:
    """Stitch all segments in a chapter into a single audio file."""
    project_path = get_current_project_path()

    if not project_path:
        raise RuntimeError("No project is currently open")

    # Check ffmpeg availability
    if not check_ffmpeg():
        raise RuntimeError("ffmpeg is not installed or not in PATH")

    # Verify chapter exists and has all audio
    status = get_chapter_audio_status(chapter_id)
    if not status.ready_to_stitch:
        raise ValueError(
            f"Chapter is not ready to stitch. Missing {status.segments_missing_audio} audio files."
        )

    audiobook_dir = get_audiobook_dir(project_path)
    segments = list_segments(chapter_id)

    # Get absolute paths for all audio files
    input_files = [str(Path(audiobook_dir) / s.audio_path) for s in segments if s.audio_path]

    # Determine output path
    import re
    safe_title = re.sub(r'[^a-zA-Z0-9]', '_', status.chapter_title)
    filename = output_filename or f"{safe_title}.mp3"
    output_dir = Path(audiobook_dir) / "exports" / "chapters"
    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / filename

    # Concatenate files
    concatenate_audio_files(input_files, str(output_path), "mp3")

    # Get actual duration of output
    duration = get_audio_duration(str(output_path))

    return StitchChapterResult(
        chapter_id=chapter_id,
        chapter_title=status.chapter_title,
        output_path=str(output_path),
        segment_count=len(segments),
        total_duration_ms=duration,
    )


def get_stitch_status() -> StitchStatus:
    """Get the overall stitch status for the book."""
    chapters = get_chapters_with_stats()

    chapter_statuses: list[ChapterStatus] = []
    for ch in chapters:
        status = get_chapter_audio_status(ch.id)
        chapter_statuses.append(ChapterStatus(
            id=ch.id,
            title=ch.title,
            sort_order=ch.sort_order,
            segment_count=ch.segment_count,
            segments_with_audio=status.segments_with_audio,
            duration_ms=status.total_duration_ms,
            ready=status.ready_to_stitch,
        ))

    total_segments = sum(ch.segment_count for ch in chapter_statuses)
    total_with_audio = sum(ch.segments_with_audio for ch in chapter_statuses)
    total_duration = sum(ch.duration_ms for ch in chapter_statuses)
    chapters_ready = sum(1 for ch in chapter_statuses if ch.ready)

    return StitchStatus(
        total_chapters=len(chapters),
        chapters_ready=chapters_ready,
        total_segments=total_segments,
        segments_with_audio=total_with_audio,
        total_duration_ms=total_duration,
        ready_to_stitch_book=chapters_ready == len(chapters) and len(chapters) > 0,
        chapters=chapter_statuses,
    )


def stitch_book(
    output_filename: Optional[str] = None,
    include_chapter_markers: bool = True,
) -> StitchBookResult:
    """Stitch all chapters into a complete audiobook."""
    project_path = get_current_project_path()

    if not project_path:
        raise RuntimeError("No project is currently open")

    # Check ffmpeg availability
    if not check_ffmpeg():
        raise RuntimeError("ffmpeg is not installed or not in PATH")

    # Check overall status
    status = get_stitch_status()
    if not status.ready_to_stitch_book:
        not_ready = [ch.title for ch in status.chapters if not ch.ready]
        raise ValueError(
            f"Book is not ready to stitch. Chapters not ready: {', '.join(not_ready)}"
        )

    audiobook_dir = get_audiobook_dir(project_path)
    info = get_project_info()
    chapters = list_chapters()

    # First, stitch each chapter and collect the chapter files
    chapter_files: list[str] = []
    chapter_markers: list[ChapterMarker] = []
    current_ms = 0

    for chapter in chapters:
        # Stitch the chapter
        result = stitch_chapter(chapter.id)
        chapter_files.append(result.output_path)

        # Record chapter marker
        chapter_markers.append(ChapterMarker(
            title=chapter.title,
            start_ms=current_ms,
        ))

        current_ms += result.total_duration_ms

    # Determine output path
    import re
    safe_title = re.sub(r'[^a-zA-Z0-9]', '_', info.project.title)
    filename = output_filename or f"{safe_title}.mp3"
    output_dir = Path(audiobook_dir) / "exports" / "book"
    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / filename

    # Create final audiobook
    if include_chapter_markers:
        metadata = {
            "title": info.project.title,
            "artist": info.project.author,
            "album": info.project.title,
        }
        create_audiobook_with_chapters(chapter_files, str(output_path), chapter_markers, metadata)
    else:
        concatenate_audio_files(chapter_files, str(output_path), "mp3")

    # Get actual duration
    duration = get_audio_duration(str(output_path))

    return StitchBookResult(
        output_path=str(output_path),
        chapter_count=len(chapters),
        total_duration_ms=duration,
        chapters=[
            ChapterTimestamp(title=cm.title, start_ms=cm.start_ms)
            for cm in chapter_markers
        ],
    )


def clear_segment_audio(segment_id: str) -> None:
    """Clear audio from a segment."""
    db = get_database()
    cursor = db.cursor()

    # Verify segment exists
    segment = get_segment(segment_id)
    if not segment:
        raise ValueError(f"Segment not found: {segment_id}")

    # Clear audio path and duration
    cursor.execute(
        "UPDATE segments SET audio_path = NULL, duration_ms = NULL WHERE id = ?",
        (segment_id,)
    )
    db.commit()
